File: src/wiki/assets.py
"""Static asset discovery, validation, and manifest helpers."""

from __future__ import annotations

import posixpath
from pathlib import Path, PurePosixPath
from urllib.parse import quote, unquote

from .config import Config
from .links import is_external_link, split_target
from .paths import OutputEntry


def iter_asset_files(config: Config) -> list[Path]:
    assets: list[Path] = []
    for asset_dir in config.wiki.assets:
        if not asset_dir.exists() or asset_dir.is_symlink():
            continue
        for path in sorted(asset_dir.rglob("*")):
            if config.is_excluded(path) or path.is_dir() or path.is_symlink():
                continue
            assets.append(path)
    return assets


def audit_assets(config: Config) -> list[str]:
    warnings: list[str] = []
    for asset_dir in config.wiki.assets:
        if config.is_excluded(asset_dir):
            continue
        if not asset_dir.exists():
            warnings.append(f"Asset directory does not exist: {asset_dir}")
        elif asset_dir.is_symlink():
            warnings.append(f"Asset directory is a symlink and will not be copied: {asset_dir}")
        elif not asset_dir.is_dir():
            warnings.append(f"Asset directory is not a directory: {asset_dir}")
        for path in sorted(asset_dir.rglob("*")) if asset_dir.exists() and asset_dir.is_dir() and not asset_dir.is_symlink() else []:
            if path.is_symlink() and not config.is_excluded(path):
                warnings.append(f"Asset symlink will not be copied: {path}")
    return warnings


_PACKAGED_ASSET_FILENAMES = frozenset()


def write_packaged_asset(filename: str, dest: Path) -> None:
    """Write a bundled asset from the wiki package to dest."""
    raise ValueError(f"Unknown packaged asset: {filename!r}")


def build_asset_manifest(config: Config, owned_output_dir: Path, base_url: str) -> list[OutputEntry]:
    entries: list[OutputEntry] = []
    base = base_url.rstrip("/") if base_url else ""
    seen_outputs: set[Path] = set()
    for asset in iter_asset_files(config):
        rel = config.relative_to_root(asset)
        rel_parts = [part for part in PurePosixPath(rel).parts if part]
        output_path = owned_output_dir.joinpath(*rel_parts)
        seen_outputs.add(output_path.resolve())
        encoded = quote(rel, safe="/()_-.$~")
        public_url = f"{base}/{encoded}" if base else f"/{encoded}"
        entries.append(OutputEntry(source=asset, output_path=output_path, public_url=public_url, kind="asset"))
    entries.extend(_packaged_asset_entries(config, owned_output_dir, base, seen_outputs))
    return entries


def _packaged_asset_entries(
    config: Config,
    owned_output_dir: Path,
    base: str,
    seen_outputs: set[Path],
) -> list[OutputEntry]:
    from importlib.resources import files as resource_files

    entries: list[OutputEntry] = []
    for filename in sorted(_PACKAGED_ASSET_FILENAMES):
        packaged = resource_files("wiki").joinpath(f"assets/{filename}")
        if not packaged.is_file():
            continue
        rel = f"assets/{filename}"
        output_path = owned_output_dir / "assets" / filename
        if output_path.resolve() in seen_outputs:
            continue
        workspace_asset = config.config_root / "assets" / filename
        source = workspace_asset if workspace_asset.is_file() else None
        encoded = quote(rel, safe="/()_-.$~")
        public_url = f"{base}/{encoded}" if base else f"/{encoded}"
        entries.append(OutputEntry(source=source, output_path=output_path, public_url=public_url, kind="asset"))
    return entries


def resolve_asset_path(config: Config, current_file: Path, target: str) -> Path | None:
    if is_external_link(target):
        return None
    page_part, _ = split_target(target)
    page_part = unquote(page_part.split("?")[0]).replace("\\", "/").strip()
    if not page_part or page_part.startswith("/"):
        return None
    try:
        current_rel = current_file.resolve().relative_to(config.config_root.resolve()).as_posix()
    except ValueError:
        current_rel = current_file.as_posix()
    current_dir = posixpath.dirname(current_rel)
    combined = posixpath.normpath(posixpath.join(current_dir, page_part))
    if combined.startswith("../") or combined == "..":
        return None
    candidate = (config.config_root / Path(combined)).resolve()
    for asset_dir in config.wiki.assets:
        try:
            candidate.relative_to(asset_dir.resolve())
            return candidate
        except ValueError:
            continue
    return None


def asset_reference_issue(config: Config, current_file: Path, target: str) -> str | None:
    asset_path = resolve_asset_path(config, current_file, target)
    if asset_path is None:
        return f"points outside configured assets: {target}"
    if config.is_excluded(asset_path):
        return f"points to excluded asset: {target}"
    if asset_path.is_symlink():
        return f"points to symlink asset, which will not be copied: {target}"
    if not asset_path.exists() or not asset_path.is_file():
        return f"points to missing asset: {target}"
    return None

File: src/wiki/assets.ts
/**
 * Static asset discovery, validation, and manifest helpers.
 *
 * Port of `src/wiki/assets.py`. Assets are the one part of a wiki that is
 * *copied* rather than rendered, so the questions this module answers are all
 * about what will and will not make it to the site:
 *
 * - which files a build copies ({@link iterAssetFiles}),
 * - which configured asset directories and symlinks it cannot copy
 *   ({@link auditAssets}),
 * - what URL each asset gets ({@link buildAssetManifest}), and
 * - whether a link that looks like an asset actually resolves to one
 *   ({@link assetReferenceIssue}) — the answer the link audit needs.
 *
 * Two Python details are load-bearing:
 *
 * - **Symlinks are skipped, not followed.** `iter_asset_files` drops a symlink
 *   and a symlinked directory, and `audit_assets` reports both, because
 *   following them can escape the config root or duplicate a tree.
 * - **Asset URLs are quoted with the wiki's own safe set** (`/()_-.$~`), the
 *   same one `paths.page_url` uses. `encodeURIComponent` would publish a
 *   different URL for a filename with parentheses, which is exactly the kind of
 *   filename an asset tends to have.
 *
 * `_PACKAGED_ASSET_FILENAMES` is an empty frozen set in the Python source, so
 * the branch that walks it never runs; the port omits it and the
 * `write_packaged_asset` raiser with it, rather than keeping unreachable code
 * that claims to do something.
 */

import {
  isDirectory,
  isFile,
  isSymlink,
  pathExists,
  relativeWithin,
  sortedTreePaths,
} from "./fspath.ts";
import { join, resolve } from "@std/path";
import type { Config } from "./config.ts";

import {
  isExternalLink,
  normalizePosixPath,
  posixDirname,
  pyUnquote,
  splitTarget,
} from "./links.ts";
import type { OutputEntry } from "./schemas/domain.ts";
import { quote } from "./urlquote.ts";

/** Every asset file a build would copy, in measured path-component order. */
export function iterAssetFiles(config: Config): string[] {
  const assets: string[] = [];
  for (const assetDir of config.wiki.assets) {
    if (!pathExists(assetDir) || isSymlink(assetDir)) continue;
    for (const path of sortedTreePaths(assetDir)) {
      if (config.isExcluded(path)) continue;
      if (isDirectory(path) || isSymlink(path)) continue;
      assets.push(path);
    }
  }
  return assets;
}

/**
 * Report the asset directories a build cannot copy from.
 *
 * This is a warning list rather than a single error because a wiki can have
 * several asset directories and one bad one should not hide the other three.
 */
export function auditAssets(config: Config): string[] {
  const warnings: string[] = [];
  for (const assetDir of config.wiki.assets) {
    if (config.isExcluded(assetDir)) continue;
    const usable = pathExists(assetDir) && isDirectory(assetDir) &&
      !isSymlink(assetDir);
    if (!pathExists(assetDir)) {
      warnings.push(`Asset directory does not exist: ${assetDir}`);
    } else if (isSymlink(assetDir)) {
      warnings.push(
        `Asset directory is a symlink and will not be copied: ${assetDir}`,
      );
    } else if (!isDirectory(assetDir)) {
      warnings.push(`Asset directory is not a directory: ${assetDir}`);
    }
    if (!usable) continue;
    for (const path of sortedTreePaths(assetDir)) {
      if (isSymlink(path) && !config.isExcluded(path)) {
        warnings.push(`Asset symlink will not be copied: ${path}`);
      }
    }
  }
  return warnings;
}

/**
 * The output entries for every asset, as a build writes them.
 *
 * The public URL is the asset's path relative to the config root, quoted with
 * the wiki's safe set and prefixed with the site's base URL — the same shape
 * `build_page_manifest` gives pages, so collision detection can compare them.
 */
export function buildAssetManifest(
  config: Config,
  ownedOutputDir: string,
  baseUrl: string,
): OutputEntry[] {
  const entries: OutputEntry[] = [];
  const base = baseUrl.replace(/\/+$/, "");
  for (const asset of iterAssetFiles(config)) {
    const rel = config.relativeToRoot(asset);
    const relParts = rel.split("/").filter((part) => part !== "");
    const outputPath = join(ownedOutputDir, ...relParts);
    const encoded = quote(rel, "/()_-.$~");
    const publicUrl = base === "" ? `/${encoded}` : `${base}/${encoded}`;
    entries.push({
      source: asset,
      output_path: outputPath,
      public_url: publicUrl,
      kind: "asset",
    });
  }
  return entries;
}

/**
 * The asset a markdown link points at, or `null` when it points at none.
 *
 * `null` is doing two jobs — "not inside an asset directory" and "escaped the
 * config root" — because the caller reports the same thing either way. A link
 * that climbs out with `../` is refused *before* resolution, so a wiki cannot
 * link into a neighbour's assets.
 */
export function resolveAssetPath(
  config: Config,
  currentFile: string,
  target: string,
): string | null {
  if (isExternalLink(target)) return null;
  const [pagePartRaw] = splitTarget(target);
  const pagePart = pyUnquote(pagePartRaw.split("?")[0] as string)
    .replaceAll("\\", "/")
    .trim();
  if (pagePart === "" || pagePart.startsWith("/")) return null;

  let currentRel: string;
  try {
    currentRel =
      (relativeWithin(resolve(currentFile), resolve(config.config_root)))
        .replaceAll("\\", "/");
  } catch {
    currentRel = currentFile.replaceAll("\\", "/");
  }
  const combined = normalizePosixPath(
    posixDirname(currentRel) === ""
      ? pagePart
      : `${posixDirname(currentRel)}/${pagePart}`,
  );
  if (combined.startsWith("../") || combined === "..") return null;

  const candidate = resolve(join(config.config_root, ...combined.split("/")));
  for (const assetDir of config.wiki.assets) {
    try {
      relativeWithin(candidate, resolve(assetDir));
      return candidate;
    } catch {
      continue;
    }
  }
  return null;
}

/**
 * Why a link to an asset is broken, or `null` when it is fine.
 *
 * The distinctions are the ones a user can act on: "outside configured assets"
 * means the asset directories are not configured for that path, "excluded"
 * means `wiki.exclude` covers it, and each has a different fix.
 */
export function assetReferenceIssue(
  config: Config,
  currentFile: string,
  target: string,
): string | null {
  const assetPath = resolveAssetPath(config, currentFile, target);
  if (assetPath === null) return `points outside configured assets: ${target}`;
  if (config.isExcluded(assetPath)) {
    return `points to excluded asset: ${target}`;
  }
  if (isSymlink(assetPath)) {
    return `points to symlink asset, which will not be copied: ${target}`;
  }
  if (!pathExists(assetPath) || !isFile(assetPath)) {
    return `points to missing asset: ${target}`;
  }
  return null;
}
